### DISCUSSION TITLE
Welche Profildatei Easymodes wurde geladen?

### DISCUSSION THREAD

**Discussion by Weseler485** [OpenCCU/OpenCCU#2805] [Development]
Hallo 🙋🏼‍♂️ 
Kann mir jemand einen Tipp geben, wie ich herausfinden kann, welche Profildatei aus den Easymodes bei einer Direktverknüpfung geladen wird?

Ich habe das DropDownMenü eines Wertes in der Direktverknüpfung um
einen Wert erweitert, der nur über den Expertenprofil eingestellt werden konnte. Funktionier auch!
Nur wird jetzt immer diese Verknüpfung im Expertenprofil angezeigt. Ich konnte schon herausfinden, dass das daran liegt, dass der Wert in keinem Profil zu finden ist und als FallBack dann das Expertenprofil herangezogen wird.

Bloß wie finde ich nun heraus, in welcher Profildatei ich den Wert ergänzen muss? Sämtliche Versuche sind bis hier hin gescheitert...

Danke euch!!

**Comment by jp112sdl** [comment]
Als grobe Richtung würde ich vermuten hier:
https://github.com/eq-3/occu/blob/825afd98b9c450bf68a0a648911f3915b671ed4e/WebUI/www/config/ic_common.tcl#L840C1-L994

Mal ein bisschen Debugausgaben einbauen und schauen, wann wie wohin eingestiegen wird und wann "0" (Expertenprofil) oder ein anderer Index zurückgegeben wird.

**Comment by Weseler485** [comment]
Danke dir für die Antwort.
Die Stelle ist soweit richtig, dass sie entscheidet, welches Profil (innerhalb einer Datei) angezeigt werden soll. Ich bekomme das Expertenprofil - und somit die Null - weil ich ein Mismatch in Zeile 974 habe. Soweit bin ich schon.

Ich bekomme es aber nicht hin herauszufinden, welche Datei in $PROFILE eingelesen wurde.

**Comment by jp112sdl** [comment]
> Ich bekomme es aber nicht hin herauszufinden, welche Datei in $PROFILE eingelesen wurde.

`/www/config/easymodes/<TARGET>/<SOURCE>.tcl` oder besser gesagt
`/www/config/easymodes/<AKTOR>/<PARTNER>.tcl`
Also bspw. du hast einen `Dimmer`-Kanal mit einer Fernbedienungs`taste` verknüpft, dann wird
`/www/config/easymodes/DIMMER/KEY.tcl` geladen.


Ist es das, was du meinst?

**Comment by Weseler485** [comment]
> > Ist es das, was du meinst?


Ja, vielen Dank! Manchmal kann die Lösung soo einfach sein!

Aber:
Im konkreten Beispiel steuere ich mit einer HmIP-RC8 (Fernbedienung, weil ich zum Test zu faul bin immer aufzustehen) die Beleuchtung eines HmIP-BSL an.
Nun hatte ich bereits irgendwie herausgefunden, dass der BSL ein DIMMER_VIRTUAL_RECEIVER ist und konnte dort das Dropdown ergänzen.

Bloß gibt es dort keine KEY.tcl sondern nur eine KEY_TRANCEIVER.tcl
Die ist aber nicht das was ich suche - da sind die falschen Profile drin :-(

kennst du noch einen Trick, wie ich von dem Gerät auf den jeweiligen "Begriff" schließen kann ohne TryAndError?

**Reply by Weseler485** [reply]
Ich korrigiere meine Feststellung. Die Datei ist doch die richtige. Ich liebe diese Altlasten, dass in der Datei eine Description hinterlegt ist, die von einer anderen Stelle dann übersteuert wird. habe mich da durch einen falschen Text verwirren lassen.

Datei geändert, funktioniert  🥳

Bleibt aber die Frage, wie ich von Geräten auf Begriffe schließen kann, ohne try and error beziehungsweise ohne jahrelange Vorkenntnisse.

**Reply by jp112sdl** [reply]
Das wird schwierig. Wenn man sich nicht (mehr) aktiv damit beschäftigt, ist das Wissen im Nu wieder raus aus der Birne.
Nach längerem Überlegen war es glaube ich auch so, dass die Geräte-XML da mitspielt.
Beispiel eines einfachen Schaltaktors: https://github.com/AskSinPP/asksinpp-web/blob/master/rftypes/xmls/rf_s.xml
```
...
      <paramset type="LINK" id="switch_ch_link">
        <subset ref="switch_linkset" />
      </paramset>
...
```

Das führt dazu, dass in der WebUI die [`/www/config/easymodes/switch_ch_link.tcl`](https://github.com/eq-3/occu/blob/master/WebUI/www/config/easymodes/switch_ch_link.tcl) geladen wird.


Und darin ist wiederum angegeben, welche Profildateien nachgeladen werden
```
...
  if {[catch {source [file join $env(DOCUMENT_ROOT) config/easymodes/$ACTOR/$x.tcl]}] != 0} {
...
```

Bei HmIP kommt man da nicht so einfach ran... zumindest steht in den XMLs in der HMIPServer.jar glaub ich das mit den Linksets so nicht mehr drin.
Aber bei deinem DIMMER-TASTER Problem würde ich vermuten, dass es die [`/www/config/easymodes/linkHmIP_DIMMER_VIRTUAL_RECEIVER.tcl`](https://github.com/eq-3/occu/blob/master/WebUI/www/config/easymodes/linkHmIP_DIMMER_VIRTUAL_RECEIVER.tcl) ist.
Und da wird dann auch wieder 
```
...
  if {[catch {source [file join $env(DOCUMENT_ROOT) config/easymodes/$ACTOR/$x.tcl]}] != 0} {
...
```
gemacht.

**Reply by Weseler485** [reply]
Nochmals danke für die ausführliche Antwort und die investierte Zeit.

Auch wenn ich es jetzt nicht ins kleinste Detail nachvollzogen habe dürftest du mit der Geräte-XML sehr gut liegen. Zumindest in der ic_ifacecmd.cgi habe ich in Zeile 241 gefunden, dass das Array dev_descr_receiver durch einen Aufruf von xmlrpc befüllt wird.

Ich werde dann mal in einiger ruhigen Minute versuchen, ob ich nicht irgendwo diese Quelle angezapft bekomme um mir den Device-Type ausgeben zu lassen.

**Reply by jp112sdl** [reply]
Ja gern, war mal wieder eine gute Übung für zwischendurch, um Wartezeiten zu überbrücken :)

**Reply by Weseler485** [reply]
So, geht zwar noch schöner - aber ich hab es :) 

![image](https://github.com/user-attachments/assets/fd99a8b7-0029-4321-9251-31c8c4e582ee)

Die benötigten Infos um die Easymodedatei zu finden, verstecken sich (zumindest innerhalb der webui.js) in der Variable channel.channelType


Wirklich vielen Dank nochmal.
Wieder was gelernt und ich kann jetzt ein paar Parameter im "normalen Profil" ergänzen ohne dass er automatisch auf das total unübersichtliche Expertenprofil umspringt.
